### ISSUE TITLE
Some issues with module bootstrapping

### ISSUE THREAD

**Issue by tgamblin** [spack/spack#5351]
It looks like #3057 introduced a bug in `setup-env.sh`.  On my mac when I source `setup-env.sh`, I now get:

```console
basename: invalid option -- 'b'
Try 'basename --help' for more information.
WARNING: A method for managing modules does not currently exist.

To resolve this you may either:
1. Allow spack to handle this by running 'spack bootstrap'
   and sourcing this script again.
2. Install and activate a supported module managment engine manually
   Supported engines include: environment-modules and lmod
$
```

I think the `basename: invalid option -- 'b'` part is a bug, and the rest is too noisy.

You can use Spack without environment modules, and once we introduce environment support, we won't *require* modules for most users on laptops/workstations.  I think `setup-env.sh` should be mostly quiet, and the warnings above should be moved to `spack load`, `spack unload`, etc., so that they trigger when you first try to load a module, not when you start your shell.

@krafczyk @scheibelp @mamelara @alalazo : opinions?

**Comment by mamelara** [comment]
I agree. Also #3057 doesn't seem to be detecting lmod installed on my mac

**Comment by tgamblin** [comment]
@mamelara: is lmod in your PATH?

**Comment by tgamblin** [comment]
Also even if you have modules installed, you get this:

```
INFO: Using spack managed module system.
$
```

Which I think should just be omitted, because users will see it on every shell login.

**Comment by scheibelp** [comment]
Originally I liked these messages because they informed the user when Spack was modifying `module` which I assumed might be of interest - `setup-env.sh` does not otherwise affect generic commands; if there is no message, IMO it probably shouldn't define `module`. I can create a PR to remove the definition from `setup-env.sh` for now.

I'm wondering if another approach is for bootstrap to optionally modify `.bashrc` as is done in http://spack.readthedocs.io/en/latest/getting_started.html:

```
TMP=`tempfile`
echo >$TMP
MODULE_HOME=`spack location --install-dir environment-modules`
MODULE_VERSION=`ls -1 $MODULE_HOME/Modules | head -1`
${MODULE_HOME}/Modules/${MODULE_VERSION}/bin/add.modules <$TMP
cp .bashrc $TMP
echo "MODULE_VERSION=${MODULE_VERSION}" > .bashrc
cat $TMP >>.bashrc
```

Would you have anything against that @krafczyk?

Also FWIW: 

> It looks like #3057 introduced a bug in setup-env.sh. On my mac when I source setup-env.sh

That should be addressed by 80ac61339164ac8044fafb641403384cb54ae3e7

**Comment by krafczyk** [comment]
Okay folks, let me try to get to all of your concerns, and let me start by saying that all of these errors seem to be OSX specific. I believe that all of these behavior specific bugs for OSX can be fixed, it just needs to be worked on a bit.

@tgamblin and @scheibelp _I do not think it is a good idea to remove the module code from `setup-env.sh`_. Let me explain why.

**Inherent ease of use and portability**

What attracted me to spack is its inherent portability and ease of use. It is one of the most light-weight package managers out there, and also one of the easiest to use and modify. This change was meant to enhance those aspects. With a change like this, the user doesn't have to do anything like modifying their `.bashrc`, spack takes care of everything. This means when you log into a new machine, you don't have to spend much time getting your shell configured, just `source setup-env.sh` and you're off. This is hugely convenient and I think anybody who uses spack would benefit from it.

**Allowing multiple spack stacks to co-exist easily**

I'm not sure about others on here, but I sometimes use multiple spack versions. This is for two reasons. First I am testing changes to spack and may be testing multiple versions simultaneously. I want be able to easily switch between them.

Second I have software which works great with an older version of spack and for which something sometimes breaks in a newer version but which I do not want fix right now. This means I must use a different spack stack for development and production.

Switching between spack stacks made easy because the only step to set up spack is `source setup-env.sh`. Just open a new terminal, and source the script of the version of spack you want to use.

@scheibelp: This is the primary reason why I don't like modifying `.bashrc` directly as a compromise. If you use multiple versions, you'll still have to do everything manually in regards to `environment-modules` if you want to switch spack stacks and your system doesn't provide `module` already.

**Noisiness and Hints for users**

Now I understand that @tgamblin wants `setup-env.sh` to be as quiet as possible. Okay, but let me be devil's advocate from two angles for a moment. For a new user to spack, I can almost guarantee you that they do not know about the need to use `environment-modules` or `lmod`. Just look at the comment thread for #3057. There is at least one person who mentioned this problem specifically, so I suspect there are definitely more of us. Even if they do know about `environment-modules`, they do not want to spend the time to set it up before giving spack a try. I know I didn't, that was the most irritating aspect of spack. This message appropriately informs the user about how spack is going to manage the environment, and what they need to do to get spack running properly as quickly as possible.

Now there is another angle to this as well. For the experienced user, it is helpful to know what spack is doing to manage the environment as well. This message tells you whether it is using the system `module` command or a spack managed `module` command. This is helpful for experienced users, because perhaps you don't want spack to manage `module` for some reason! This lets you know right away.

@tgamblin, I get that silence on the command line is nice, however my feeling when nothing is written to the terminal is that something went wrong. I have no idea whether the command completed successfully.  The only way to find out is by executing `$?`, and if the code is 0 then it's successful. This was especially annoying when creating this PR because at one point I had introduced a bug to `setup-env.sh` so that it was exiting early by virtue of the `-e` option in the shebang and printing no output. The only clue was that tests were mysteriously failing. In fact, I'd like to advocate for a nice message to say that the environment is properly setup at the end.

**Addressing bugs**

I'm embarrassed that there are so many bugs in this PR. I'm thinking about what I can do in the future to prevent my PRs from containing so many bugs. I feel as if the bugs may be detracting from what this PR is trying to accomplish.

@tgamblin Have you tried the command with the fix that @scheibelp mentioned? This should take care of the `basename` error.

@tgamblin, The intent of this code is to say `INFO: Using system available module system.` in the event that `module` or `use` is already a shell function. So this sounds like a bug which needs fixing. Which machine are you on when you get this message?

@mamelara The code I added detects the existence of the `use` function for lmod. It's possible that my detection algorithm is failing for `use`. I'll see whether I can get lmod working on my machine to test it.

**Conclusions**

- These changes make spack more portable and easy to use.
- These changes help new users get spack running as quickly as possible
- These changes help experienced users detect environment problems more quickly
- It is not good to have a quiet setup script.
- There **are** some bugs here which need fixing.

Let me know your thoughts.

**Comment by mamelara** [comment]
Our consultants are not liking the noise either here at NERSC. We are also getting: 
```INFO: Using system available module system.```. I guess there is concern that it may break scp (according to a consultant). 

Also I do not have lmod in my PATH but rather I have the module commands in my bash_profile. I'm relatively new to installing and using lmod on my mac so I may have done this incorrectly.

**Comment by krafczyk** [comment]
Why would it break scp? Can you get an explanation from this consultant? As far as I know scp is only used for copying files..

Can you execute these commands for me?
`type module`
`type use`

**Comment by mamelara** [comment]
```
➜  ~ type module
module is a shell function from /Users/mamelara/.zshrc
➜  ~ type use
use not found
```

**Comment by scheibelp** [comment]
> This is the primary reason why I don't like modifying .bashrc directly as a compromise. If you use multiple versions, you'll still have to do everything manually in regards to environment-modules if you want to switch spack stacks and your system doesn't provide module already.

That makes sense

> [Printing messages] is helpful for experienced users, because perhaps you don't want spack to manage module for some reason! This lets you know right away.

That was my thought on why it shouldn't define `module` if it was quiet. There may be an alternative though: perhaps `setup-env.sh` could take a command line option which tells it to try defining `module` using Spack-installed `environment-modules`. I think it would be fine if it were quiet in that case.

Overall I think the original intent is that `setup-env.sh` should be quiet (which I missed when reviewing your PR), so if this happens in `setup-env.sh` it should happen quietly; doing it optionally would make that OK IMO. Another option is to add another script called `use-spack-environments`.

**Comment by scheibelp** [comment]
Overall I'm not against adding functionality which has the effect of #3057, but the impression I'm getting from this thread is that it would be best to try it another way. I think we could temporarily resolve this conflict by merging https://github.com/LLNL/spack/pull/5352 and then trying again. Do folks like any of the suggestions in https://github.com/LLNL/spack/issues/5351#issuecomment-329239782?

**Comment by krafczyk** [comment]
@mamelara It looks like you're using `environment-modules` and not `lmod`, in which case `module` should already be defined, and this code should say `INFO: Using system available module system.`.

@scheibelp: If people are not convinced by my arguments above that `setup-env.sh` should not be totally silent, Then my favorite compromise option is passing the command line argument.

**Comment by alalazo** [comment]
> I think setup-env.sh should be mostly quiet, and the warnings above should be moved to spack load, spack unload, etc., so that they trigger when you first try to load a module, not when you start your shell.

I agree with this. This would raise a warning exactly when a module system is needed. Other than that, in general I prefer a successful command to be silent, rather than verbose (following the usual Unix style). The only place where I want things the opposite way is in commands that could do unrecoverable actions (delete or overwrite something). There I normally appreciate to have flags to turn off verbosity and user interaction, but I also like that those things are on if I don't specify otherwise.

Concerning @krafczyk post above, I agree with almost every point. The only thing that doesn't convince me is noisiness. I think we shouldn't be echoing anything when we source `setup_env.sh`, but apart from that the current behavior is quite convenient.

> Why would it break scp? Can you get an explanation from this consultant? As far as I know scp is only used for copying files..

I think they may refer to the behavior reported in this [bug report](https://bugzilla.redhat.com/show_bug.cgi?id=20527) which, as far as I know, still applies. In that respect breaking `.bashrc` for some services just because you sourced `setup_env.sh` in there is in my opinion more severe than breaking it because you executed `spack load` in it.

> I'm embarrassed that there are so many bugs in this PR. I'm thinking about what I can do in the future to prevent my PRs from containing so many bugs. I feel as if the bugs may be detracting from what this PR is trying to accomplish.

@krafczyk Please don't be! I don't think this PR is bad, or extremely buggy, it just has some details that needs to be polished. Probably the only part people don't like is the echoing in `setup_env.sh`, but I personally find all the other additions useful.

To be fair, I think there's another issue not related to the PR that could raise the level of concerns for bugs in the setup phase of Spack: *we don't release often and features are still moving rather quickly*. While the latter is a good sign, the two things together mean that many people are using `develop` as their production environment and are updating it regularly to get new features. This is just to say: I am pretty sure that if we had more frequent releases an occasional bug in `develop` wouldn't be as important as it is now.

**Comment by tgamblin** [comment]
> > I'm embarrassed that there are so many bugs in this PR. I'm thinking about what I can do in the future to prevent my PRs from containing so many bugs. I feel as if the bugs may be detracting from what this PR is trying to accomplish.
> 
> @krafczyk Please don't be! I don't think this PR is bad, or extremely buggy, it just has some details that needs to be polished. Probably the only part people don't like is the echoing in setup_env.sh, but I personally find all the other additions useful.

@krafczyk: I haven't had time to get through the entire discussion here, but I completely agree with @alalazo's post, particularly the part above.  Also we reviewed your PR 😄 .  So don't feel too bad.  Welcome to OSS!  If you feel bad about your PR, feel free to go through and find all the bugs I've caused in Spack.  I can assure you I've caused far more.

**Comment by krafczyk** [comment]
@alalazo and @tgamblin: thanks for the kind words!

Presumably, if we can make sure that the `module` and `use` command is detected correctly for all users, Would folks be amenable to just removing the echo statements? Then when using spack load, it could check whether `module` or `use` is defined and complain otherwise, giving the warning I defined above. I'm dissapointed to learn about this scp bug which @alalazo mentions above, and this means the quiet folks are probably right for now.

This would be okay in my opinion. You still get the benefits of having `module` automatically defined for you in `setup-env.sh` if you need it, but with an appropriate message from a different part of spack telling you what to do if things aren't configured correctly.

**Comment by krafczyk** [comment]
#5365 Does part of the job I mention above. I would work on a second commit to implement the `spack load` side of things.

**Comment by alalazo** [comment]
#5377 is another request related to the changes in #3057

**Comment by tgamblin** [comment]
I do think it would be good if lmod were the default module system.  It does require more dependencies but it's the only maintained module system.

**Comment by alalazo** [comment]
@tgamblin What if we let a user choose which module system they want loaded through an entry in `config.yaml`?

**Comment by tgamblin** [comment]
Could we just get that from the order of enabled module systems?  I.e., prefer the first one?  Then we wouldn't need a new option.

**Comment by alalazo** [comment]
We ship some defaults for both (at least in #3183) in `etc/spack/defaults`. I don't know how that interacts with the order of keys in the merged `modules.yaml` if different files in the config hierarchy have a different relative ordering of the same options.

**Comment by tgamblin** [comment]
Ok `module_roots` is a `dict`, so no order there, but modules:enable in modules.yaml is a list so why not use the first thing listed there?

**Comment by krafczyk** [comment]
@tgamblin @alalazo I've submited #5955 To add `lmod` and `dotkit` support using @tgamblin's method. Please have a look.

**Comment by alalazo** [comment]
The main issue seems to be fixed in #5365. Closing this issue, but feel free to reopen if I missed any point.
